### DISCUSSION TITLE
is there a filesystem for USB mass storage?

### DISCUSSION THREAD

**Discussion by WaveSimm** [micropython/micropython#14383] [m.iMX RT / Teensy 4.x] [answered]
I want to use external usb memory for Teensy4.1.
I read the Teensy4.1 has a USB host port and related library for MIDI or keyboard.
However I can't find the library for USB Mass Storagy..
Is it impossible?

**Comment by jonnor** [comment, marked as answer]
USB host is not currently support in MicroPython. On any port. USB device code was just merged, and will be in MicroPython 1.23.

**Reply by WaveSimm** [reply]
Hi.. I have got 1.23 version.. Is it possible to use USB external memory now? If then, how can I get the information to do that?

**Reply by robert-hh** [reply]
The USB device basic functions are supported now an a few ports. Now someone has to write the code for an USB MSC device. The MIMXRT port has not yet support for USB devices. That may come soon.

**Reply by andrewleech** [reply]
If anyone has time to look into writing a USB storage driver i _highly_ recommend trying for MTP (media transport protocol) rather than Mass Storage. 

MTP is used by phones etc to provide file transfers without the risk of disk corruption that plagues mass storage on embedded devices.
It's built into windows and Linux at least. Apple has refused to support it, probably as a protest against Android phone popularity. There's third party apps and drivers for it now though.
